### PR TITLE
emit checkpoints on Envelope construction and invocation

### PR DESCRIPTION
**Pull request by richardstartin**
This tracks akka envelopes, enforces exclusion of mailbox scheduling and prevents tracing of a `ForkJoinTask` internal to `Dispatcher`

### CODE REVIEW

File: dd-java-agent/instrumentation/akka-concurrent/src/akka23Test/groovy/AkkaActorTest.groovy [hunk 1]
```diff
@@ -81,4 +87,41 @@ class AkkaActorTest extends AgentTestRunner {
...
+    def barrier = akkaTester.block(name)
+    def threadMigrations = threadMigrationsPerWorkflow * n
+    when:
+    runUnderTrace("parent") {
+      (1..n).each {i -> akkaTester.send(name, "who $i")}
+    }
+    barrier.release()
+    // FIXME the expected interaction counts are stable with a pause
+    //  but it should be possible to synchronise this test better
+    Thread.sleep(1000)
```
Comment (richardstartin) [inline dd-java-agent/instrumentation/akka-concurrent/src/akka23Test/groovy/AkkaActorTest.groovy:102]: @bantonsson how do I do this better? I thought the barrier would help.

File: dd-java-agent/instrumentation/akka-concurrent/src/akka23Test/groovy/AkkaActorTest.groovy [hunk 2]
```diff
@@ -81,4 +87,36 @@ class AkkaActorTest extends AgentTestRunner {
...
+      (1..n).each {i -> akkaTester.send(name, "who $i")}
+    }
+    barrier.release()
+    TEST_WRITER.waitForTraces(1)
+    then:
+    (2 * n + 1) * TEST_CHECKPOINTER.checkpoint(_, _, SPAN)
+    n * envelopes * TEST_CHECKPOINTER.checkpoint(_, _, THREAD_MIGRATION)
+    n * envelopes * TEST_CHECKPOINTER.checkpoint(_, _, THREAD_MIGRATION | END)
+    n * envelopes * TEST_CHECKPOINTER.checkpoint(_, _, CPU | END)
+    (2 * n + 1) * TEST_CHECKPOINTER.checkpoint(_, _, SPAN | END)
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/akka-concurrent/src/akka23Test/groovy/AkkaActorTest.groovy:105]: Will this test fail if there are more interactions than expected?

File: dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinExecutorTaskInstrumentation.java [hunk 3]
```diff
@@ -60,12 +60,7 @@ public void adviceTransformations(AdviceTransformation transformation) {
   public static final class Construct {
     @Advice.OnMethodExit
     public static void construct(@Advice.Argument(0) Runnable wrapped) {
-      TraceScope activeScope = activeScope();
-      if (null != activeScope) {
-        InstrumentationContext.get(Runnable.class, State.class)
-            .putIfAbsent(wrapped, State.FACTORY)
-            .captureAndSetContinuation(activeScope);
-      }
+      capture(InstrumentationContext.get(Runnable.class, State.class), wrapped, true);
```
Comment (bantonsson) [inline dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinExecutorTaskInstrumentation.java:63]: Love the cleanup...

**Review by bantonsson — APPROVED** [review]
👍

File: dd-java-agent/instrumentation/akka-concurrent/src/akka23Test/groovy/AkkaActorTest.groovy (hunk 2 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/akka-concurrent/src/akka23Test/groovy/AkkaActorTest.groovy:105]: yes

File: dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinTaskInstrumentation.java [hunk 4]
```diff
@@ -73,35 +85,26 @@ public void adviceTransformations(AdviceTransformation transformation) {
...
     public static <T> TraceScope before(@Advice.This ForkJoinTask<T> task) {
-      return AdviceUtils.startTaskScope(
-          InstrumentationContext.get(ForkJoinTask.class, State.class), task);
+      return startTaskScope(InstrumentationContext.get(ForkJoinTask.class, State.class), task);
     }
 
     @Advice.OnMethodExit(onThrowable = Throwable.class)
-    public static void after(@Advice.Enter TraceScope scope) {
-      AdviceUtils.endTaskScope(scope);
+    public static void after(@Advice.Enter TraceScope scope, @Advice.This ForkJoinTask<?> task) {
```
Comment (mcculls) [inline dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinTaskInstrumentation.java:92]: `@Advice.This ForkJoinTask<?> task` argument looks unused and could be removed?

**Review by mcculls — APPROVED** [review]

File: dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinTaskInstrumentation.java (hunk 4 above)
Comment (richardstartin) [reply dd-java-agent/instrumentation/akka-concurrent/src/main/java/datadog/trace/instrumentation/akka/concurrent/AkkaForkJoinTaskInstrumentation.java:92]: Thanks, debugging artifact

**Pull request merged**